proposals/NNNN-implicit-self-explicit-capture.md

```diff
@@ -0,0 +1,189 @@
+
+In codebases that make heavy use of asynchronous code, such as clients of [PromiseKit](https://github.com/mxcl/PromiseKit), or even Apple's new [Combine](https://developer.apple.com/documentation/combine) and [SwiftUI](https://developer.apple.com/documentation/swiftui) libraries, maintainers must either choose to adopt style guides which require the use of explicit `self` in all cases, or else resign to the reality that explicit usage of `self` will appear inconsistently throughout the code. Given that Apple's preferred/recommended style is to omit `self` where possible (as evidenced by examples throughout [The Swift Programming Language](https://docs.swift.org/swift-book/) and the [SwiftUI Tutorials](https://developer.apple.com/tutorials/swiftui)), having any asynchronous code littered with `self.` is a suboptimal state of affairs.
+
+The error is also overly conservative—it will prevent the usage of implicit `self` even when it is very unlikely that the capture of `self` will somehow cause a reference cycle. With function builders in SwiftUI, the automatic resolution to the "make capture semantics explicit" error (and indeed, the fix-it that is currently supplied to the programmer) is just to slap a `self.` on wherever the compiler tells you to. While this likely fine when `self` has value semantics, building this habit could cause users to ignore the error and apply the fix-it in cases where it is not in fact appropriate.
+
+## Proposed solution
+
+Allow the use of implicit `self` when it appears in the closure's capture list. The above code could then be written as:
+
+```swift
+execute { [self] in
+    let foo = doFirstThing()
+    performWork(with: bar)
+    doSecondThing(with: foo)
+    cleanup()
+}
+```
+
+This change still forces code which captures `self` to be explicit about its intentions, but reduces the cost of that explicitness to a single declaration. With this change explicit capture of `self` would be one of *two* ways to get rid of the error, with the current method of adding `self.` to each property/method access (without adding `self` to the capture list) remaining as a viable option.
+
+The compiler would also offer an additional fix-it when implicit `self` is used:
+
+```swift
+execute { // <- Fix-it: capture 'self' explicitly to enable implicit 'self' in this closure. Fix-it: insert '[self] in'
+    let foo = doFirstThing()
+    performWork(with: bar)
+    doSecondThing(with: foo)
+    cleanup()
+}
+```
+
+Furthermore, if the type of `self` has value semanics, we will not require any explicit usage of `self` (at the call/use site or in the capture list), so that if `self` were a `struct` or `enum` then the above could be written as simply:
```

Comment:
```suggestion
Second, if the type of `self` has value semanics, we will not require any explicit usage of `self` (at the call/use site or in the capture list), so that if `self` were a `struct` or `enum` then the above could be written as simply:
```